joomla/joomla-cms
# adding Sub_part info to mySQLi db import/export code

Allows for export and import of keys with length values

To test:

Run this sql query to create a test table

```
CREATE TABLE `dbkeytest` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
   `fileupload` text,
  PRIMARY KEY (`id`),
  KEY `fb_join_fileupload_INDEX` (`fileupload`(10)) )
```

In index.php after:
```
$app = JFactory::getApplication('site');
```

add:

```
jimport('joomla.filesystem.file');
$res =  (string) JFactory::getDbo()->getExporter()->from('dbkeytest');
JFile::write(JPATH_SITE . '/tmp/dbkeytest.xml', $res);
echo $res; exit;
```

refresh your site and you should see a white page, view the page source and you should see. 

```

<?xml version="1.0"?>
<mysqldump xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
 <database name="">
  <table_structure name="dbkeytest">
   <field Field="id" Type="int(11)" Null="NO" Key="PRI" Extra="auto_increment" />
   <field Field="fileupload" Type="text" Null="YES" Key="MUL" Extra="" />
   <key Table="dbkeytest" Non_unique="0" Key_name="PRIMARY" Seq_in_index="1" Column_name="id" Collation="A" Sub_part ="" Null="" Index_type="BTREE" Comment="" />
   <key Table="dbkeytest" Non_unique="1" Key_name="fb_join_fileupload_INDEX" Seq_in_index="1" Column_name="fileupload" Collation="A" Sub_part ="10" Null="YES" Index_type="BTREE" Comment="" />
  </table_structure>
 </database>
</mysqldump>
```
An xml file should also have been saved in /tmp/dbkeytest.xml.

Note that Sub_part ="10" has been added to the key 'fb_join_fileupload_INDEX'

The importer will also work *IF* you have this PR merged in as well 
https://issues.joomla.org/tracker/joomla-cms/7378

To test:

Drop the table dbkeytest

Replace the test code you added to index.php with:

```
JFactory::getDbo()->getImporter()->from(file_get_contents(JPATH_SITE . '/tmp/dbkeytest.xml'))
		->asXml()->mergeStructure();
exit;
```

Refresh the browser and you should see that the db table dbkeytest has been re-created in the database and that the key  'fb_join_fileupload_INDEX' has the sub_part applied. ie if you run:

```
show create table dbkeytest;
```
the results should be

```
CREATE TABLE `dbkeytest` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `fileupload` text,
  PRIMARY KEY (`id`),
  KEY `fb_join_fileupload_INDEX` (`fileupload`(10))
) ENGINE=InnoDB DEFAULT CHARSET=utf8
```
